### PR TITLE
RS/YJ/Rule 11-6 (Re)

### PR DESCRIPTION
**Pull request by yunjoonjung-PNNL**
I re-created this 11-6 PR because I had some issue with the old branch. I copied/pasted Weili's comments from the old PR.

### CODE REVIEW

File: rct229/rulesets/ashrae9012019/section11/section11rule6.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
+from rct229.rule_engine.rule_base import RuleDefinitionBase
```
Comment (yunjoonjung-PNNL) [inline rct229/rulesets/ashrae9012019/section11/section11rule6.py:1]: Weili's comment:

My take is this rule should be at ```service_water_heating_distribution_systems``` level, not building level. (See RDS Evaluation Context)

In RMD Rule, we need to add an ```is_applicable``` to check if there are ```service_water_heating_distribution_systems``` or not.
In RMD Rule ```create_data```, we should call ```get_swh_components_associated_with_each_swh_distribution_system``` to get the data.

In the ```ServiceWaterHeatingDistributionSystem```, we need to check every piping, including children, is modeled with thermal losses turned to false.
You can reuse the same code block searches for piping under a ```ServiceWaterHeatingDistributionSystem``` from the ```get_swh_components_associated_with_each_swh_distribution_system``` function.

File: rct229/rulesets/ashrae9012019/section11/section11rule6.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    class RMDRule(RuleDefinitionListIndexedBase):
+        def __init__(self):
+            super(Section11Rule6.RMDRule, self).__init__(
+                rmds_used=produce_ruleset_model_description(
+                    USER=False, BASELINE_0=True, PROPOSED=False
+                ),
+                each_rule=Section11Rule6.RMDRule.SWHDistRule(),
+                index_rmd=BASELINE_0,
+                list_path="$.service_water_heating_distribution_systems[*]",
+            )
```
Comment (yunjoonjung-PNNL) [inline rct229/rulesets/ashrae9012019/section11/section11rule6.py:40]: Need a ```is_applicable``` here.

File: rct229/rulesets/ashrae9012019/section11/section11rule6.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+            return swh_dist_sys_b
+
+        def create_data(self, context, data):
+            rmd_b = context.BASELINE_0
+
+            bldg_segment_ids = find_all("$.buildings[*].building_segments[*].id", rmd_b)
+
+            for bldg_segment_id in bldg_segment_ids:
+                swh_distribution_and_eq_list_b = (
+                    get_swh_uses_associated_with_each_building_segment(
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section11/section11rule6.py:58]: As we discussed - this is incorrect, the return type of `get_swh_uses_associated_with_each_building_segment` is either a list of `swh_use` associate with the provided `building_segment` or a dict of `building_segment_id` to a list of `swh_use`. It is not `swh_distribution`.

File: rct229/rulesets/ashrae9012019/section11/section11rule6.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+            )
+
+            return swh_dist_sys_b
+
+        def create_data(self, context, data):
+            rmd_b = context.BASELINE_0
+
+            bldg_segment_ids = find_all("$.buildings[*].building_segments[*].id", rmd_b)
+
+            for bldg_segment_id in bldg_segment_ids:
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section11/section11rule6.py:56]: This for loop is confusing too - it is just keep overriding the `swh_distribution_and_eq_list_b`.

**Review by weilixu — COMMENTED** [review]
This implementation needs update since the function `get_swh_uses_associated_with_each_building_segment`.

File: rct229/rulesets/ashrae9012019/section11/section11rule6.py (hunk 3 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule6.py:58]: Good point. Addressed with the ` a dict of building_segment_id to a list of swh_use` suggestion.

File: rct229/rulesets/ashrae9012019/section11/section11rule6.py (hunk 4 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule6.py:56]: Addressed.

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        for swh_use_id in find_all(
+            f'$.buildings[*].building_segments[*][?(@.id="{building_segment["id"]}")].zones[*].spaces[*].service_water_heating_uses[*].id',
+            rmd,
+        ):
+            swh_use = find_exactly_one_with_field_value(
+                f"$.zones[*].spaces[*].service_water_heating_uses[*]",
+                "id",
+                swh_use_id,
+                building_segment,
+            )
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:57]: Addressed

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+                swh_use_dict[area_type] += sum(swh_use_energy_by_space.values())
+            else:
+                for space_id in swh_use_energy_by_space:
+                    if space_id != "no_spaces_assigned":
+                        space = find_exactly_one_with_field_value(
+                            "$.zones[*].spaces[*]",
+                            "id",
+                            space_id,
+                            building_segment,
+                        )
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:81]: good catch! Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule6.py [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+                get_swh_equipment_associated_with_each_swh_distribution_system(rmd_b)
+            )
+
+            return {"swh_distribution_and_eq_list_b": swh_distribution_and_eq_list_b}
+
+        def list_filter(self, context, data=None):
+            swh_dist_sys_b = context.BASELINE_0
+            swh_distribution_and_eq_list_b = data["swh_distribution_and_eq_list_b"]
+
+            return swh_dist_sys_b["id"] in swh_distribution_and_eq_list_b
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule6.py:62]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule12.py [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+            def is_applicable(self, context, data=None):
+                building_b = context.BASELINE_0
+                building_p = context.PROPOSED
+                schedules_b = data["schedules_b"]
+                is_leap_year_b = data["is_leap_year_b"]
+
+                service_water_heating_uses_p = find_all(
+                    "$.building_segments[*].zones[*].spaces[*].service_water_heating_uses[*]",
+                    building_p,
+                )
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule12.py:82]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule12.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+                building_open_schedule_p = find_one(
+                    "$.building_open_schedule", building_p
+                )
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule12.py:85]: Addressed.

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+                                service_water_heating_bat
+                            ] += swh_use_energy_by_space[space_id]
+                        else:
+                            swh_use_dict.setdefault("UNDETERMINED", ZERO.ENERGY)
+                            swh_use_dict["UNDETERMINED"] += swh_use_energy_by_space[
+                                space_id
+                            ]
+
+        building_segment_swh_bat = (
+            max(swh_use_dict, key=swh_use_dict.get) if swh_use_dict else None
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:99]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule6.py [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                    },
+                )
+
+            def get_calc_vals(self, context, data=None):
+                swh_dist_sys_b = context.BASELINE_0
+
+                piping_losses_modeled_b = [
+                    service_water_piping.get("are_thermal_losses_modeled")
+                    for service_water_piping in swh_dist_sys_b["service_water_piping"]
+                ]
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule6.py:83]: Addressed.

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+                "id",
+                swh_use_id,
+                building_segment,
+            )
+
+            if (
+                not swh_use
+                or swh_use.get("use_units") == SERVICE_WATER_HEATING_USE_UNIT.OTHER
+            ):
+                return None
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:63]: Addressed.

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py [hunk 13]
```diff
@@ -0,0 +1,103 @@
...
+                "swh_distribution": [],
+                "swh_heating_eq": [],
+                "pumps": [],
+                "tanks": [],
+                "piping": [],
+                "solar_thermal": [],
+                "swh_uses": [],
+                "energy_required": ZERO.ENERGY,
+            },
+        )
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:47]: Good point. Addressed.

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+
+        for swh_use_id in find_all(
+            f'$.buildings[*].building_segments[*][?(@.id="{building_segment["id"]}")].zones[*].spaces[*].service_water_heating_uses[*].id',
+            rmd,
+        ):
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:52]: Addressed.

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+            swh_use = find_exactly_one_service_water_heating_use(rmd, swh_use_id)
+            energy_required = get_energy_required_to_heat_swh_use(
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:54]: Addressed.

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+                swh_use_id, rmd, building_segment["id"], is_leap_year
+            )
+            if list(energy_required.values())[0] is not None:
+                swh_and_equip_dict[swh_bat]["energy_required"] += list(
+                    energy_required.values()
+                )[0]
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_components_associated_with_each_swh_bat.py:60]: @Jiarongx-Xie Could you answer this question?

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py [hunk 17]
```diff
@@ -0,0 +1,102 @@
...
+    building_segment_id: str
+        building segment id
+    is_leap_year: bool, default: False
+        Whether the year is a leap year or not.
+
+    Returns
+    -------
+    building_segment_swh_bat: str
+        one of the ServiceWaterHeatingSpaceOptions2019ASHRAE901 options
+
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/ruleset_functions/get_building_segment_swh_bat.py:32]: Addressed.

File: rct229/rulesets/ashrae9012019/section11/section11rule12.py [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+
+            def get_calc_vals(self, context, data=None):
+                swh_uses_associated_with_each_building_segment_p = data[
+                    "swh_uses_associated_with_each_building_segment_p"
+                ]
+
+                return {
+                    "swh_uses_associated_with_each_building_segment_p": swh_uses_associated_with_each_building_segment_p
+                }
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section11/section11rule12.py:119]: Addressed.

**Comment by yunjoonjung-PNNL** [comment]
This PR is closed due to complex branch management. Each rule/function will have a separate PR including only relevant codes.

File: rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_uses_associated_with_each_building_segment.py [hunk 19]
```diff
@@ -2,27 +2,28 @@
...
         RMD at RuleSetModelDescription level
-    building_segment_id: str
-        building segment id
 
     Returns
     -------
-    swh_uses_list: list
-        A list containing the ids of all service water heating uses associated with a building segment
+    swh_uses_dict: dict
+        A dictionary where the keys are all the building segment ids and the values are from the `use` key under the `service_water_heating_uses`.
```
Comment (Jiarongx-Xie) [inline rct229/rulesets/ashrae9012019/ruleset_functions/get_swh_uses_associated_with_each_building_segment.py:19]: Are the values the swh use object or the field "use"?

**Pull request closed without merging**